horovod/tensorflow/__init__.py

```diff
@@ -215,6 +215,123 @@ def allreduce_grads(grads):
         return allreduce_grads
 
 
+class LocalGradientAggregationHelper:
+    def __init__(self, aggregation_frequency, allreduce_func=None):
+        if allreduce_func == None:
+            allreduce_func = lambda x: x
+        self._allreduce_grads = allreduce_func
+
+        # How often are parameters synchronized
+        self.aggregation_frequency = aggregation_frequency
+        assert self.aggregation_frequency > 0
+
+        # This is going to be N data structure holding the aggregated gradient updates
+        # for parameter updates. N is the number of parameters.
+        self.gpu_shadow_vars = []
+
+        # Used to know when to allreduce and apply gradients. We allreduce when `self.counter`
+        # is equal to `self.aggregation_frequency`. We apply gradients when `self.counter` is
+        # equal to 0.
+        self.counter = None
+
+    def init_aggregation_vars(self, grads):
+        with tf.variable_scope("aggregation_variables"):
+            self.counter = tf.get_variable(
+                "aggregation_counter", shape=(), dtype=tf.int32,
+                trainable=False, initializer=tf.zeros_initializer())
+            if self.aggregation_frequency > 1:
+                for idx, grad in enumerate(grads):
+                    grad_aggregation_variable_name = str(idx)
+                    grad_aggregation_variable = tf.get_variable(
+                        grad_aggregation_variable_name, shape=grad.get_shape().as_list(),
+                        trainable=False, initializer=tf.zeros_initializer(), dtype=tf.float64,
+                        collections=[tf.GraphKeys.LOCAL_VARIABLES, "aggregating_collection"])
+                    self.gpu_shadow_vars.append(grad_aggregation_variable)
+                assert len(self.gpu_shadow_vars) == len(grads)
+
+    def clear_grads(self):
+        clear_ops_list = []
+        for idx, grad in enumerate(self.gpu_shadow_vars):
+            grad_aggregation_variable_name = str(idx)
+            grad_aggregator = tf.get_variable(
+                grad_aggregation_variable_name, dtype=tf.float64)
```

Comment:
oof yeah -- so I got an error running the mnist example where it was using tf.float64 but the default is tf.float32

maybe what we need to do is make the `dtype` equal to whatever the `dtype` of `grad` is?

Reply:
Yeah we should definitely be using the `dtype` of `grad` if possible, and i don't see why it wouldn't be possible. Because it's possible we are communicating way too much if we are rounding everything up to `float64`. 

I am not opposed to using a dict, it's though considered better tensorflow practice to use `tf.get_variable`. I think we should be able to get rid of `dtype` in all the but the initial call which _should_ make this simpler.